File: native/cocos/renderer/pipeline/PlanarShadowQueue.cpp
/****************************************************************************
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "PlanarShadowQueue.h"
#include "Define.h"
#include "InstancedBuffer.h"
#include "PipelineSceneData.h"
#include "PipelineStateManager.h"
#include "RenderInstancedQueue.h"
#include "RenderPipeline.h"
#include "core/geometry/AABB.h"
#include "gfx-base/GFXDevice.h"
#include "scene/Camera.h"
#include "scene/Model.h"
#include "scene/RenderScene.h"
#include "scene/Shadow.h"

namespace cc {
namespace pipeline {

PlanarShadowQueue::PlanarShadowQueue(RenderPipeline *pipeline)
: _pipeline(pipeline),
  _phaseID(getPhaseID("planar-shadow")) {
    _instancedQueue = ccnew RenderInstancedQueue;
}

PlanarShadowQueue::~PlanarShadowQueue() {
    destroy();
}

void PlanarShadowQueue::gatherShadowPasses(scene::Camera *camera, gfx::CommandBuffer *cmdBuffer) {
    clear();

    const PipelineSceneData *sceneData = _pipeline->getPipelineSceneData();
    scene::Shadows *shadowInfo = sceneData->getShadows();
    if (shadowInfo == nullptr || !shadowInfo->isEnabled() || shadowInfo->getType() != scene::ShadowType::PLANAR || shadowInfo->getNormal().length() < 0.000001F) {
        return;
    }

    const auto *scene = camera->getScene();
    const bool shadowVisible = camera->getVisibility() & static_cast<uint32_t>(LayerList::DEFAULT);
    if (!scene->getMainLight() || !shadowVisible) {
        return;
    }

    const auto &models = scene->getModels();
    for (const auto &model : models) {
        if (scene->isCulledByLod(camera, model)) {
            continue;
        }
        if (!model->isEnabled() || !model->isCastShadow() || !model->getNode()) {
            continue;
        }

        if (model->getWorldBounds() && model->isCastShadow()) {
            _castModels.emplace_back(model);
        }
    }

    geometry::AABB ab;
    for (const auto *model : _castModels) {
        // frustum culling
        model->getWorldBounds()->transform(shadowInfo->getMatLight(), &ab);
        if (!ab.aabbFrustum(camera->getFrustum())) {
            continue;
        }

        const auto &subModels = model->getSubModels();
        for (const auto &subModel : subModels) {
            const auto shadowPassIdx = getShadowPassIndex(subModel);
            if (shadowPassIdx == -1) {
                _subModelArray.emplace_back(subModel);
                _shaderArray.emplace_back(shadowInfo->getPlanarShader(subModel->getPatches()));
                _passArray.emplace_back((*shadowInfo->getMaterial()->getPasses())[0]);
                continue;
            }
            auto *const pass = subModel->getPass(shadowPassIdx);
            const auto batchingScheme = pass->getBatchingScheme();
            if (batchingScheme == scene::BatchingSchemes::INSTANCING) {
                auto *instancedBuffer = pass->getInstancedBuffer();
                instancedBuffer->merge(subModel, shadowPassIdx);
                _instancedQueue->add(instancedBuffer);
            } else { // standard draw
                _subModelArray.emplace_back(subModel);
                _shaderArray.emplace_back(subModel->getShader(shadowPassIdx));
                _passArray.emplace_back(pass);
            }
        }
    }

    _instancedQueue->uploadBuffers(cmdBuffer);
}

void PlanarShadowQueue::clear() {
    _castModels.clear();
    _subModelArray.clear();
    _shaderArray.clear();
    _passArray.clear();
    if (_instancedQueue) _instancedQueue->clear();
}

void PlanarShadowQueue::recordCommandBuffer(gfx::Device *device, gfx::RenderPass *renderPass, gfx::CommandBuffer *cmdBuffer, uint32_t subpassID) {
    const PipelineSceneData *sceneData = _pipeline->getPipelineSceneData();
    const auto *shadowInfo = sceneData->getShadows();
    if (shadowInfo == nullptr || !shadowInfo->isEnabled() || shadowInfo->getType() != scene::ShadowType::PLANAR || shadowInfo->getNormal().length() < 0.000001F) {
        return;
    }

    _instancedQueue->recordCommandBuffer(device, renderPass, cmdBuffer);

    for (size_t i = 0; i < _subModelArray.size(); ++i) {
        const auto *const subModel = _subModelArray[i];
        auto *const shader = _shaderArray[i];
        const auto *pass = _passArray[i];
        auto *const ia = subModel->getInputAssembler();
        auto *const pso = PipelineStateManager::getOrCreatePipelineState(pass, shader, ia, renderPass, subpassID);

        cmdBuffer->bindPipelineState(pso);
        cmdBuffer->bindDescriptorSet(materialSet, pass->getDescriptorSet());
        cmdBuffer->bindDescriptorSet(localSet, subModel->getDescriptorSet());
        cmdBuffer->bindInputAssembler(ia);
        cmdBuffer->draw(ia);
    }
}

void PlanarShadowQueue::destroy() {
    _pipeline = nullptr;
    CC_SAFE_DELETE(_instancedQueue);
    _castModels.clear();
    _subModelArray.clear();
}

int PlanarShadowQueue::getShadowPassIndex(const scene::SubModel *subModel) const {
    int i = 0;
    for (const auto &pass : *(subModel->getPasses())) {
        if (pass->getPhase() == _phaseID) {
            return i;
        }
        ++i;
    }
    return -1;
}

} // namespace pipeline
} // namespace cc

File: cocos/rendering/planar-shadow-queue.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { cclegacy, geometry } from '../core';
import { isEnableEffect, SetIndex } from './define';
import { CommandBuffer, Device, RenderPass, Shader } from '../gfx';
import { PipelineStateManager } from './pipeline-state-manager';
import { Model, Camera, SubModel } from '../render-scene/scene';
import { RenderInstancedQueue } from './render-instanced-queue';
import { ShadowType } from '../render-scene/scene/shadows';
import { Layers } from '../scene-graph/layers';
import { PipelineRuntime } from './custom/pipeline';
import { BatchingSchemes, Pass } from '../render-scene/core/pass';
import { getPhaseID } from './pass-phase';

const _ab = new geometry.AABB();
let _phaseID = getPhaseID('planar-shadow');
function getPlanarShadowPassIndex (subModel: SubModel): number {
    const passes = subModel.passes;
    const r = cclegacy.rendering;
    if (isEnableEffect()) _phaseID = r.getPhaseID(r.getPassID('default'), 'planar-shadow');
    for (let k = 0; k < passes.length; k++) {
        if (((!r || !r.enableEffectImport) && passes[k].phase === _phaseID)
        || (isEnableEffect() && passes[k].phaseID === _phaseID)) {
            return k;
        }
    }
    return -1;
}

export class PlanarShadowQueue {
    private _subModelArray: SubModel[] = [];
    private _shaderArray: Shader[] = [];
    private _passArray: Pass[] = [];
    private _castModels: Model[] = [];
    private _instancedQueue = new RenderInstancedQueue();
    private _pipeline: PipelineRuntime;

    constructor (pipeline: PipelineRuntime) {
        this._pipeline = pipeline;
    }

    /**
     * @en
     * clear planar-shadow queue
     * @zh
     * 清除 planar-shadow 渲染队列数据
     */
    public clear (): void {
        this._subModelArray.length = 0;
        this._shaderArray.length = 0;
        this._passArray.length = 0;
        this._instancedQueue.clear();
        this._castModels.length = 0;
    }

    public gatherShadowPasses (camera: Camera, cmdBuff: CommandBuffer): void {
        this.clear();
        const pipelineSceneData = this._pipeline.pipelineSceneData;
        const shadows = pipelineSceneData.shadows;
        if (!shadows.enabled || shadows.type !== ShadowType.Planar || shadows.normal.length() < 0.000001) { return; }

        const scene = camera.scene!;
        const frustum = camera.frustum;
        const shadowVisible =  (camera.visibility & Layers.BitMask.DEFAULT) !== 0;
        if (!scene.mainLight || !shadowVisible) { return; }

        const models = scene.models;
        const visibility = camera.visibility;
        for (let i = 0; i < models.length; i++) {
            const model = models[i];
            if (scene.isCulledByLod(camera, model)) {
                continue;
            }
            if (model.enabled && model.node && model.castShadow
                && (model.node && ((visibility & model.node.layer) === model.node.layer))) {
                this._castModels.push(model);
            }
        }

        for (let i = 0; i < this._castModels.length; i++) {
            const model = this._castModels[i];
            if (model.worldBounds) {
                geometry.AABB.transform(_ab, model.worldBounds, shadows.matLight);
                if (!geometry.intersect.aabbFrustum(_ab, frustum)) { continue; }
            }

            const subModels = model.subModels;
            for (let j = 0; j < subModels.length; j++) {
                const subModel = subModels[j];
                const shadowPassIdx = getPlanarShadowPassIndex(subModel);
                if (shadowPassIdx < 0) {
                    this._subModelArray.push(subModel);
                    const planarShader = shadows.getPlanarShader(subModel.patches);
                    if (!planarShader) { continue; }
                    this._shaderArray.push(planarShader);
                    this._passArray.push(shadows.material.passes[0]);
                    continue;
                }

                const pass = subModel.passes[shadowPassIdx];
                const batchingScheme = pass.batchingScheme;
                if (batchingScheme === BatchingSchemes.INSTANCING) {
                    const buffer = pass.getInstancedBuffer();
                    buffer.merge(subModel, shadowPassIdx);
                    this._instancedQueue.queue.add(buffer);
                } else {
                    const shader = subModel.shaders[shadowPassIdx];
                    this._subModelArray.push(subModel);
                    if (shader) this._shaderArray.push(shader);
                    this._passArray.push(pass);
                }
            }
        }
        this._instancedQueue.uploadBuffers(cmdBuff);
    }

    public recordCommandBuffer (device: Device, renderPass: RenderPass, cmdBuff: CommandBuffer): void {
        const shadows = this._pipeline.pipelineSceneData.shadows;

        if (!shadows.enabled || shadows.type !== ShadowType.Planar) { return; }
        this._instancedQueue.recordCommandBuffer(device, renderPass, cmdBuff);

        for (let i = 0; i < this._subModelArray.length; ++i) {
            const subModel = this._subModelArray[i];
            const shader = this._shaderArray[i];
            const pass = this._passArray[i];
            const ia = subModel.inputAssembler;
            const pso = PipelineStateManager.getOrCreatePipelineState(device, pass, shader, renderPass, ia);
            const descriptorSet = pass.descriptorSet;

            cmdBuff.bindPipelineState(pso);
            cmdBuff.bindDescriptorSet(SetIndex.MATERIAL, descriptorSet);
            cmdBuff.bindDescriptorSet(SetIndex.LOCAL, subModel.descriptorSet);
            cmdBuff.bindInputAssembler(ia);
            cmdBuff.draw(ia);
        }
    }
}
